### PR TITLE
VCF convert + VEP

### PR DESCRIPTION
**Pull request by BertGalle**
## PR checklist

- [ ] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/nf-cmgg/exomecnv/tree/master/.github/CONTRIBUTING.md)
- [ ] Make sure your code lints (`nf-core lint`).
- [ ] Ensure the test suite passes (`nf-test test main.nf.test -profile test,docker`).
- [ ] Check for unexpected warnings in debug mode (`nextflow run . -profile debug,test,docker --outdir <OUTDIR>`).
- [ ] Usage Documentation in `docs/usage.md` is updated.
- [ ] Output Documentation in `docs/output.md` is updated.
- [ ] `CHANGELOG.md` is updated.
- [ ] `README.md` is updated (including new tool citations and authors/contributors).

### CODE REVIEW

File: modules/local/exomedepth/merge_cnv/main.nf
Comment (nvnieuwk) [inline modules/local/exomedepth/merge_cnv/main.nf:1]: Can you add a container to this and conda declaration to this module? It could otherwise cause issues when the pipeline would ever be run on cloud. 

Can you also add some versioning to this module for reproducibility sake? (like in the other modules)

File: modules/local/exomedepth/merge_cnv/main.nf [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+    publishDir "$params.outdir/exomedepth/cnv_call", mode: 'copy'
+
+    input:
+    tuple val(meta), path(auto), path(chrx)
+
+    output:
+    tuple val(meta), path("*.txt")
+
+    script:
+    prefix = task.ext.prefix ?: "${meta}_CNVs_ExomeDepth"
```
Comment (nvnieuwk) [inline modules/local/exomedepth/merge_cnv/main.nf:14]:
```suggestion
    def prefix = task.ext.prefix ?: "${meta}_CNVs_ExomeDepth"
```

This makes sure this variable can only be accessed in this `script` block

File: nextflow_schema.json [hunk 2]
```diff
@@ -292,24 +296,58 @@
             "description": "Options specific for the ExomeDepth execution",
             "default": "",
             "properties": {
+                "yamlconfig": {
+                    "type": "string",
+                    "description": "YAML config file to use in BEDGOVCF module"
```
Comment (nvnieuwk) [inline nextflow_schema.json:301]:
```suggestion
                    "description": "YAML config file to use in BEDGOVCF module",
                    "exists": true,
                    "format": "file-path"
```
Some extra validation parameters :)

File: nextflow_schema.json [hunk 3]
```diff
@@ -292,24 +296,58 @@
...
+                },
                 "roi_auto": {
                     "type": "string",
                     "description": "Path to the default ROI (regions of interest) BED file of autosomal (merged) regions to be used for CNV analysis",
                     "format": "file-path",
                     "pattern": "^\\S+\\.bed$",
-                    "mimetype": "text/plain",
-                    "default": null
+                    "mimetype": "text/plain"
```
Comment (nvnieuwk) [inline nextflow_schema.json:308]:
```suggestion
                    "mimetype": "text/plain",
                    "exists": true
```

File: nextflow_schema.json [hunk 4]
```diff
@@ -292,24 +296,58 @@
...
                 },
                 "roi_chrx": {
                     "type": "string",
                     "description": "Path to the default ROI (regions of interest) BED file of chrX (merged) regions to be used for CNV analysis",
                     "format": "file-path",
                     "pattern": "^\\S+\\.bed$",
-                    "mimetype": "text/plain",
-                    "default": null
+                    "mimetype": "text/plain"
```
Comment (nvnieuwk) [inline nextflow_schema.json:315]:
```suggestion
                    "mimetype": "text/plain",
                    "exists": true
```

File: nextflow_schema.json [hunk 5]
```diff
@@ -292,24 +296,58 @@
...
+                },
+                "species": {
+                    "type": "string",
+                    "description": "Species",
+                    "default": "homo_sapiens"
+                },
+                "vep_cache": {
+                    "type": "string",
+                    "description": "Path to the VEP cache"
+                },
```
Comment (nvnieuwk) [inline nextflow_schema.json:339]:
```suggestion
                "vep_cache": {
                    "type": "string",
                    "format": "file-path",
                    "description": "Path to the VEP cache"
                },
```
You can also add `exists` here if you change the `vep_cache` for your tests to a real directory (doesn't need to be a VEP cache, just a real directory)

File: nextflow_schema.json [hunk 6]
```diff
@@ -292,24 +296,58 @@
...
+                "annotate": {
+                    "type": "string",
+                    "description": "Path to VCF file to run VEP on"
+                },
```
Comment (nvnieuwk) [inline nextflow_schema.json:343]: I think it would be better to add this VCF to the samplesheet. This will make it possible to run the annotation for more than one VCF at a time

File: subworkflows/local/exomedepth/main.nf
Comment (nvnieuwk) [inline subworkflows/local/exomedepth/main.nf:1]: Why did you move the code from the main workflow to a subworkflow (just curious it's not wrong)

**Review by nvnieuwk — CHANGES_REQUESTED** [review]
Hi good job! Here are some suggestions

File: modules/local/exomedepth/merge_cnv/main.nf
Comment (BertGalle) [reply modules/local/exomedepth/merge_cnv/main.nf:1]: Added for both merge modules

File: modules/local/exomedepth/merge_cnv/main.nf (hunk 1 above)
Comment (BertGalle) [reply modules/local/exomedepth/merge_cnv/main.nf:14]: Fixed

File: nextflow_schema.json (hunk 2 above)
Comment (BertGalle) [reply nextflow_schema.json:301]: Added

File: nextflow_schema.json (hunk 3 above)
Comment (BertGalle) [reply nextflow_schema.json:308]: Added

File: nextflow_schema.json (hunk 4 above)
Comment (BertGalle) [reply nextflow_schema.json:315]: Added

File: nextflow_schema.json (hunk 5 above)
Comment (BertGalle) [reply nextflow_schema.json:339]: Added

File: nextflow_schema.json (hunk 6 above)
Comment (BertGalle) [reply nextflow_schema.json:343]: Description was poorly formulated. It is the path to the samplesheet that contains a column with the VCF files. It will run annotation for more than one VCF at a time. I fixed the description.

File: subworkflows/local/exomedepth/main.nf
Comment (BertGalle) [reply subworkflows/local/exomedepth/main.nf:1]: In the future, there may be more tools added so ExomeDepth is already put in a subworkflow.

File: modules/local/exomedepth/merge_cnv/main.nf [hunk 7]
```diff
@@ -2,24 +2,40 @@
 process CNV_MERGE {
     tag "$meta"
 
+    container "quay.io/biocontainers/coreutils:9.3"
```
Comment (nvnieuwk) [inline modules/local/exomedepth/merge_cnv/main.nf:5]:
```suggestion
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://depot.galaxyproject.org/singularity/coreutils:9.3' :
        'biocontainers/coreutils:9.3' }"
```

This makes sure that the singularity containers is used in singularity instead of having to convert the docker containers to singularity (which can cause issues sometimes)

File: modules/local/exomedepth/merge_cnv/main.nf [hunk 8]
```diff
@@ -2,24 +2,40 @@
...
-    prefix = task.ext.prefix ?: "${meta}_CNVs_ExomeDepth"
+    def prefix = task.ext.prefix ?: "${meta}_CNVs_ExomeDepth"
     """
     cp $auto "${prefix}.txt"
     tail +2 $chrx >> "${prefix}.txt"
+
+    cat <<-END_VERSIONS > versions.yml
+    "${task.process}":
+        tail:\$(tail --version | sed '1!d; s/tail//')
+        cp:\$(cp --version | sed '1!d; s/cp//')
```
Comment (nvnieuwk) [inline modules/local/exomedepth/merge_cnv/main.nf:26]:
```suggestion
        tail:\$(tail --version | sed '1!d; s/tail (GNU coreutils) //')
        cp:\$(cp --version | sed '1!d; s/cp (GNU coreutils) //')
```
This way only the version is in the `versions.yml` file

File: modules/local/exomedepth/merge_cnv/main.nf [hunk 9]
```diff
@@ -2,24 +2,40 @@
...
 
     stub:
     def prefix = task.ext.prefix ?: "${meta}_CNVs_ExomeDepth"
     """
     touch ${prefix}.txt
+
+    cat <<-END_VERSIONS > versions.yml
+    "${task.process}":
+        tail:\$(tail --version | sed '1!d; s/tail//')
+        cp:\$(cp --version | sed '1!d; s/cp//')
```
Comment (nvnieuwk) [inline modules/local/exomedepth/merge_cnv/main.nf:38]:
```suggestion
        tail:\$(tail --version | sed '1!d; s/tail (GNU coreutils) //')
        cp:\$(cp --version | sed '1!d; s/cp (GNU coreutils) //')
```

File: modules/local/exomedepth/merge_count/main.nf
Comment (nvnieuwk) [inline modules/local/exomedepth/merge_count/main.nf:1]: Can you apply the same suggestions from `merge_cnv` to this module?

**Review by nvnieuwk — CHANGES_REQUESTED** [review]
Some small things left, but almost there!

**Review by nvnieuwk — COMMENTED** [review]
Looking very good! Can you also add support for bgzipped VCF files as input to the annotation? (This should be easy as it just means you should tweak the `pattern` in the schema and add an additional schema value for the VCF index (`.vcf.gz.tbi`))

As far as I know, the pipeline actually already generates bgzipped VCFs using the `bedgovcf` module, right?

**Review by nvnieuwk — APPROVED** [review]
Looks good to me! :rocket: @ToonRosseel feel free to merge this!

**Comment by ToonRosseel** [comment]
Well done @BertGalle !
